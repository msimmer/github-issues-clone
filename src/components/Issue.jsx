// @flow

import type { Issue as IssueType } from "../types";

import * as React from "react";
import Labels from "./Labels";
import Assignees from "./Assignees";
import { diffTimeToHuman } from "../util";

const Issue = (props: IssueType): React.Element<"div"> => (
  <div className="issue">
    <div className="issue__icon material-icons">
      {props.state === "OPEN" ? "error_outline" : "check_circle_outline"}
    </div>
    <div className="issue__details">
      <div className="issue__details__title">
        <div className="issue__details__title--text">{props.title}</div>
        <Labels {...props.labels} />
      </div>
      <div className="issue__details__meta">
        <div>#{props.number}</div>
        <div>opened {diffTimeToHuman(props.publishedAt)} ago by</div>
        {props.author && props.author.login ? (
          <div>{props.author.login}</div>
        ) : null}
      </div>
    </div>
    <div className="issue__assignees">
      <Assignees {...props.assignees} />
    </div>
    <div className="issue__comments">
      {props.comments.totalCount > 0 ? (
        <div className="issue__comments__counter">
          <div className="material-icons">chat_bubble_outline</div>
          <div>{props.comments.totalCount}</div>
        </div>
      ) : (
        ""
      )}
    </div>
  </div>
);

export default Issue;
