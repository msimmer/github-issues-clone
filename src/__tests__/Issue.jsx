import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { configure } from "enzyme";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import Issue from "../components/Issue";

configure({ adapter: new Adapter() });

describe("<Issue />", () => {
  const RealDate = Date;
  const publishedAt = 1549641976000;
  const now = 1549643215436;

  const mockProps = {
    state: "OPEN",
    title: "foo",
    labels: [],
    number: 1,
    publishedAt,
    author: {
      login: "bar"
    },
    assignees: [],
    comments: {
      totalCount: 0
    }
  };

  function mockDate() {
    global.Date = class extends RealDate {
      constructor() {
        return new RealDate(publishedAt);
      }
      static now() {
        return now;
      }
    };
  }

  afterEach(() => (global.Date = RealDate));
  beforeEach(() => mockDate());

  it("matches the snapshots", () => {
    const component = shallow(<Issue {...mockProps} />);
    expect(toJSON(component)).toMatchSnapshot();

    component.setProps({ labels: [{ name: "l1", color: "000000" }] });
    expect(toJSON(component)).toMatchSnapshot();

    component.setProps({ comments: { totalCount: 1 } });
    expect(toJSON(component)).toMatchSnapshot();
  });

  it("renders the date", () => {
    const component = shallow(<Issue {...mockProps} />);
    expect(component.find(".issue__details__meta").text()).toBe(
      "#1opened 21 minutes ago bybar"
    );
  });

  it("renders the icons", () => {
    const component = shallow(<Issue {...mockProps} />);
    expect(component.find(".issue__icon").text()).toBe("error_outline");

    component.setProps({ state: "CLOSE" });
    expect(component.find(".issue__icon").text()).toBe("check_circle_outline");
  });
});
