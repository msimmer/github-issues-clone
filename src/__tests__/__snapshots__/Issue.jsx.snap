// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Issue /> matches the snapshots 1`] = `
<div
  className="issue"
>
  <div
    className="issue__icon material-icons"
  >
    error_outline
  </div>
  <div
    className="issue__details"
  >
    <div
      className="issue__details__title"
    >
      <div
        className="issue__details__title--text"
      >
        foo
      </div>
      <Labels />
    </div>
    <div
      className="issue__details__meta"
    >
      <div>
        #
        1
      </div>
      <div>
        opened 
        21 minutes
         ago by
      </div>
      <div>
        bar
      </div>
    </div>
  </div>
  <div
    className="issue__assignees"
  >
    <Assignees />
  </div>
  <div
    className="issue__comments"
  />
</div>
`;

exports[`<Issue /> matches the snapshots 2`] = `
<div
  className="issue"
>
  <div
    className="issue__icon material-icons"
  >
    error_outline
  </div>
  <div
    className="issue__details"
  >
    <div
      className="issue__details__title"
    >
      <div
        className="issue__details__title--text"
      >
        foo
      </div>
      <Labels
        0={
          Object {
            "color": "000000",
            "name": "l1",
          }
        }
      />
    </div>
    <div
      className="issue__details__meta"
    >
      <div>
        #
        1
      </div>
      <div>
        opened 
        21 minutes
         ago by
      </div>
      <div>
        bar
      </div>
    </div>
  </div>
  <div
    className="issue__assignees"
  >
    <Assignees />
  </div>
  <div
    className="issue__comments"
  />
</div>
`;

exports[`<Issue /> matches the snapshots 3`] = `
<div
  className="issue"
>
  <div
    className="issue__icon material-icons"
  >
    error_outline
  </div>
  <div
    className="issue__details"
  >
    <div
      className="issue__details__title"
    >
      <div
        className="issue__details__title--text"
      >
        foo
      </div>
      <Labels
        0={
          Object {
            "color": "000000",
            "name": "l1",
          }
        }
      />
    </div>
    <div
      className="issue__details__meta"
    >
      <div>
        #
        1
      </div>
      <div>
        opened 
        21 minutes
         ago by
      </div>
      <div>
        bar
      </div>
    </div>
  </div>
  <div
    className="issue__assignees"
  >
    <Assignees />
  </div>
  <div
    className="issue__comments"
  >
    <div
      className="issue__comments__counter"
    >
      <div
        className="material-icons"
      >
        chat_bubble_outline
      </div>
      <div>
        1
      </div>
    </div>
  </div>
</div>
`;
